### PR TITLE
first draft of transformations module

### PR DESCRIPTION
**Pull request by Juliamg**
## Description
Please describe the change you have made.

## Checklist:
- [ ] Tests added/updated.
- [ ] Documentation updated.
- [ ] Changelog updated in [CHANGELOG.md](https://github.com/cognitedata/power-ops-sdk/blob/master/CHANGELOG.md).
- [ ] Version bumped. If triggering a new release is desired, bump the version number in [pyproject.toml](https://github.com/cognitedata/power-ops-sdk/blob/master/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

**Comment by Juliamg** [comment]
* Some transformations have dynamic attributes that need to be provided at runtime in preprocessor. 
* apply method as generic so that it can be run like this:
`for transformation in mapping.transformations:
                datapoints = transformation.apply(
                    time_series_data=datapoints)`
This however, requires that some transformations have setter or preprocessing steps that sets the dynamic attributes before apply is run. Can filter on the transformation name and do that step before apply. 
This is true for the transformations:
- AddWaterInTransit
- StaticValues
- HeightToVolume

File: cognite/powerops/transformations/transformations.py [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+
+#TODO: utility to test, remove later
+def t(hour: int, minutes: int = 0) -> pd.Timestamp:
+    """Simplify creation of datetimes"""
+    return pd.Timestamp(2000, 1, 1, hour, minutes)
+
+
+if __name__ == "__main__":
+
+    ### A small test script ###
```
Comment (doctrino) [inline cognite/powerops/transformations/transformations.py:352]: Why not make a unit test?

File: cognite/powerops/transformations/transformations.py [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+            start_time + timedelta(minutes=int(dp.offset_minute) + shift_minutes) for dp in relative_datapoints
+        ],
+    )
+
+
+class Transformation(BaseModel, abc.ABC):
+    transformation_name: ClassVar[str]
+
+    @classmethod
+    def load(cls, transformation_: dict[str, Any]) -> "Transformation":
```
Comment (doctrino) [inline cognite/powerops/transformations/transformations.py:55]:
```suggestion
    def load(cls, transformation_: dict[str, Any]) -> Self:
```
`from typing_extensions import Self`

File: cognite/powerops/transformations/transformations.py [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+    @property
+    def volumes(self) -> list[float]:
+        return self._volumes
+
+    @property
+    def heights(self) -> list[float]:
+        return self._heights
+
+    # TODO: run this function in preprocessor before running apply. Need a check on transformation name
+    def set_and_prepare_volumes(self, model: dict, object_type: str, instance: str | int):
```
Comment (doctrino) [inline cognite/powerops/transformations/transformations.py:144]: I think maybe you should have an optional `init` method in the base class to cover this case.

File: cognite/powerops/transformations/transformations.py [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+
+    def apply(self, time_series_data: pd.Series) -> pd.Series:
+        if self.volumes and self.heights:
+            return self.height_to_volume(time_series_data, self.heights, self.volumes)
+        else:
+            raise ValueError(f"Volumes and Heights properties had not "
+                             f"been set for transformation {self.transformation_name}.")
+
+
+class DoNothing(Transformation):
```
Comment (doctrino) [inline cognite/powerops/transformations/transformations.py:166]: Do you know the use case for this?

File: cognite/powerops/transformations/transformations.py [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+                  "offset_value": 1
+                - "offset_minute": 60
+                  "offset_value": 0.5
+    - "StaticValues":
+          "input":
+            "relative_datapoints":
+                - "offset_minute": 0
+                  "offset_value": 1
+    - "ToBool":
+          "input":
```
Comment (doctrino) [inline cognite/powerops/transformations/transformations.py:380]: Maybe make `input` optional?

**Review by doctrino — COMMENTED** [review]
Just a few comments while I looked over it.

File: cognite/powerops/transformations/transformations.py [hunk 6]
```diff
@@ -203,92 +189,63 @@ def apply(self, time_series_data: pd.Series) -> pd.Series:
...
-    def discharge(self):
-        return self._discharge
+class AddWaterInTransit(DynamicTransformation):
+    discharge_ts_external_id: str
+    transit_object_type: Literal["plant", "gate"]
+    transit_object_name: str
+    shape: Optional[dict[int, float]] = None
+    discharge: Optional[pd.Series] = None
+    start: Optional[int] = None
+    end: Optional[int] = None
```
Comment (doctrino) [inline cognite/powerops/transformations/transformations.py:199]: Good use case for pydantic private attributes?
https://docs.pydantic.dev/latest/concepts/models/#private-model-attributes

**Review by doctrino — COMMENTED** [review]
Just a few minor things, then ready to merge.

File: cognite/powerops/transformations/transformations.py [hunk 7]
```diff
@@ -39,111 +39,97 @@ def _relative_datapoints_to_series(
...
 
-    @abc.abstractmethod
-    def apply(self,
-              time_series_data: pd.Series,
-              ):
+    @abstractmethod
+    def apply(
+        self,
+        time_series_data: pd.Series,
+    ):
```
[the remark this answers is not in the document]
Comment (Juliamg) [reply cognite/powerops/transformations/transformations.py:69]: it is the black hook from pre-commit that does that

File: cognite/powerops/transformations/transformations.py [hunk 8]
```diff
@@ -39,111 +39,97 @@ def _relative_datapoints_to_series(
...
         (transformation_name,) = transformation_
-        class_index = [cls.__name__ for cls in Transformation.__subclasses__()].index(transformation_name)
-        if transformation_input := transformation_[transformation_name]["input"]:
-            return Transformation.__subclasses__()[class_index].__call__(**transformation_input)
+        subclasses = list(
+            chain(
+                [sub for cls in Transformation.__subclasses__() for sub in cls.__subclasses__()],
+                Transformation.__subclasses__(),
+            )
+        )
```
[the remark this answers is not in the document]
Comment (Juliamg) [reply cognite/powerops/transformations/transformations.py]: see the function for creating this dict, and then lookup by the name

File: cognite/powerops/transformations/transformations.py [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+        if self.pre_apply_has_run:
+            return self.add_water_in_transit(
+                inflow=time_series_data,
+                discharge=self.discharge,
+                shape=self.shape,
+                start=ms_to_datetime(self.start),
+                end=ms_to_datetime(self.end),
+            )
+        else:
+            raise ValueError("pre_apply function has not run - missing neccessary properties to run transformation")
```
Comment (doctrino) [inline cognite/powerops/transformations/transformations.py:362]:
```suggestion
            raise ValueError("pre_apply function has not run - missing neccessary properties to run transformation")
            
           
_TRANSFORMATIONS_BY_CLASS_NAME = dict(
        inspect.getmembers(
            sys.modules[__name__], lambda member: inspect.isclass(member) and member.__module__ == __name__
        )
    )
```

Comment (doctrino) [reply cognite/powerops/transformations/transformations.py:362]: Og så bruke denne over.

**Review by doctrino — APPROVED** [review]

**Pull request merged**